opentracing/scope_managers/flask.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) The OpenTracing Authors.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+
+import threading
+
+from flask import _request_ctx_stack as stack
+
+from opentracing import Scope, ScopeManager
+from .constants import ACTIVE_ATTR
+
+
+class FlaskScopeManager(ScopeManager):
+    """
+    :class:`~opentracing.ScopeManager` implementation for **Flask**
+    that stores the :class:`~opentracing.Scope` in a Flask
+    :class:`RequestContext`, made accessible via
+    :attr:`flask._request_ctx_stack`.
+    """
```

Comment:
- Why is this needed?  
- When would I want to use that instead of normal thread-local ScopeManager? What benefits does this one provide?
- We have a flask project under contrib, what's the reason not to have this defined there?

Reply:
@yurishkuro, though I can't speak from direct experience of tls not being adequate, according to the [Werkzeug docs](http://werkzeug.pocoo.org/docs/0.14/local/) thread-local storage isn't reliable in all use cases/configurations and using the provided mechanism that guarantees request-local storage seems sound for instrumented Flask apps:
```
[...] Also, whether every request gets its own thread is not guaranteed in WSGI. It could be that a request is reusing a thread from a previous request, and hence data is left over in the thread local object.
```

As to the location of the scope manager, I was keeping in line with what appears to be the current practice of including them in this repo (like Tornado's).  I'd be happy to move it to contrib's project.

Reply:
I'm not sure if the gevent scope manager is also subject to similar issues, but the werkzeug context locals seem to provide request-local storage for it as well.